src/Storages/PartitionStrategy.h

```diff
@@ -0,0 +1,119 @@
+#pragma once
+
+#include <Interpreters/ExpressionActions.h>
+#include <Parsers/IAST_fwd.h>
+
+#include <Processors/Chunk.h>
+
+namespace DB
+{
+
+/*
+ * Class responsible for computing and generating a partition key for object storage.
+ * As of now, there are two possible implementations: hive and stringfied.
+ *
+ * It also offers some helper APIs like `getFormatChunk` and `getFormatHeader`. Required mostly because of `hive` strategy
+ * since the default behavior is not to write partition columns in the files and rely only on the filepath.
+ */
+struct PartitionStrategy
+{
+    struct PartitionExpressionActionsAndColumnName
+    {
+        ExpressionActionsPtr actions;
+        std::string column_name;
+    };
+
+    PartitionStrategy(ASTPtr partition_by_, const Block & sample_block_, ContextPtr context_);
+
+    virtual ~PartitionStrategy() = default;
+
+    virtual ColumnPtr computePartitionKey(const Chunk & chunk) = 0;
+
+    virtual std::string getPathForRead(const std::string & prefix) = 0;
+    virtual std::string getPathForWrite(const std::string & prefix, const std::string & partition_key) = 0;
+
+    virtual Chunk getFormatChunk(const Chunk & chunk) { return chunk.clone(); }
+
+    virtual Block getFormatHeader() { return sample_block; }
+
+    const NamesAndTypesList & getPartitionColumns() const;
+
+protected:
+    ASTPtr partition_by;
+    Block sample_block;
+    ContextPtr context;
+    NamesAndTypesList partition_columns;
+};
+
+/*
+ * Tries to create a partition strategy given a strategy name.
+ * Performs validation on required arguments by each strategy. Example: Partition strategy hive can not be used without a PARTITION BY expression
+ */
+struct PartitionStrategyFactory
+{
+    enum class StrategyType
+    {
+        WILDCARD,
+        HIVE
```

Comment:
There is an inconsistency that here we have `wildcard` and `hive` strategy types, but according classes are called `stringified strategy` and `hive strategy`. It makes the code less understandable, let's make it consistent.